runtime/parachains/src/paras_inherent.rs

```diff
@@ -682,28 +682,52 @@ where
 fn random_sel<X, F: Fn(&X) -> Weight>(
 	rng: &mut rand_chacha::ChaChaRng,
 	selectables: Vec<X>,
+	mut preferred_indices: Vec<usize>,
 	weight_fn: F,
 	weight_limit: Weight,
 ) -> (Weight, Vec<usize>) {
 	if selectables.is_empty() {
 		return (0 as Weight, Vec::new())
 	}
-	let mut indices = (0..selectables.len()).into_iter().collect::<Vec<_>>();
+	// all indices that are not part of the preferred set
+	let mut indices = (0..selectables.len())
+		.into_iter()
+		.filter(|idx| !preferred_indices.contains(idx))
+		.collect::<Vec<_>>();
 	let mut picked_indices = Vec::with_capacity(selectables.len().saturating_sub(1));
```

Comment:
Where is that saturating_sub coming from? Why do we expect `picked_indices` to be smaller by 1?

Reply:
Because the `fn` is only every called if _not all_ of the items fit into the bin :)

Reply:
Hah, ok that makes sense.